codegen/smithy-aws-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/aws/protocols/Rpcv2Cbor.kt

```diff
@@ -0,0 +1,150 @@
+            smithyProtocolHeaderMiddleware,
+            eventStreamsAcceptHeaderMiddleware,
+            businessMetricsMiddleware,
+        )
+    }
+
+    /**
+     * Exact copy of [AwsHttpBindingProtocolGenerator.renderSerializeHttpBody] but with a custom
+     * [OperationShape.hasHttpBody] function to handle protocol-specific serialization rules.
+     */
+    override fun renderSerializeHttpBody(
+        ctx: ProtocolGenerator.GenerationContext,
+        op: OperationShape,
+        writer: KotlinWriter,
+    ) {
+        val resolver = getProtocolHttpBindingResolver(ctx.model, ctx.service)
+        if (!op.hasHttpBody(ctx)) return
+
+        // payload member(s)
+        val requestBindings = resolver.requestBindings(op)
+        val httpPayload = requestBindings.firstOrNull { it.location == HttpBinding.Location.PAYLOAD }
+        if (httpPayload != null) {
+            renderExplicitHttpPayloadSerializer(ctx, httpPayload, writer)
+        } else {
+            val documentMembers = requestBindings.filterDocumentBoundMembers()
+            // Unbound document members that should be serialized into the document format for the protocol.
+            // delegate to the generate operation body serializer function
+            val sdg = structuredDataSerializer(ctx)
+            val opBodySerializerFn = sdg.operationSerializer(ctx, op, documentMembers)
+            writer.write("val payload = #T(context, input)", opBodySerializerFn)
+            writer.write("builder.body = #T.fromBytes(payload)", RuntimeTypes.Http.HttpBody)
+        }
+        renderContentTypeHeader(ctx, op, writer, resolver)
+    }
+
+    /**
+     * @return whether the operation input does _not_ target the unit shape ([UnitTypeTrait.UNIT])
+     */
+    private fun OperationShape.hasHttpBody(ctx: ProtocolGenerator.GenerationContext): Boolean {
+        val input = ctx.model.expectShape(inputShape).targetOrSelf(ctx.model).let {
+            // If the input has been synthetically cloned from the original (most likely),
+            // pull the archetype and check _that_
+            it.getTrait<SyntheticClone>()?.let { clone ->
+                ctx.model.expectShape(clone.archetype).targetOrSelf(ctx.model)
+            } ?: it
+        }
+
+        return input.id != UnitTypeTrait.UNIT
+    }
+
+    override fun renderContentTypeHeader(
+        ctx: ProtocolGenerator.GenerationContext,
+        op: OperationShape,
+        writer: KotlinWriter,
+        resolver: HttpBindingResolver,
+    ) {
+        writer.write("builder.headers.setMissing(\"Content-Type\", #S)", resolver.determineRequestContentType(op))
+    }
+
+    class Rpcv2CborHttpBindingResolver(
+        model: Model,
+        val serviceShape: ServiceShape,
+    ) : StaticHttpBindingResolver(
+        model,
+        serviceShape,
+        HttpTrait.builder().code(200).method("POST").uri(UriPattern.parse("/")).build(),
+        "application/cbor",
+        TimestampFormatTrait.Format.EPOCH_SECONDS,
+    ) {
```

Comment:
It's not used, it could be any value, `TimestampFormatTrait.Format.UNKNOWN` might be better. This is related to your previous comment about unused timestamp format but it will be tricky to untangle